Windows calc engine integration

vidanovic opened the pull request:
# Window Calculation Engine Module
EnergyPlus contains several window calculation modules that are based on same standard and produce very similar results. In addition to this, LBNL have separate software (WINDOW and OPTICS) that is specialized in calculation of window's thermal and optical properties. WINDOW engine is constantly enhanced and debugged, and it would benefit to have window calculation engine separated into external module that can be easily updated with new algorithms.
Window Calculation Engine (WCE) is consolidation of thermal and optical routines from EnergyPlus, WINDOW and Optics into separate components that can now be used independently (https://github.com/LBNL-ETA/Windows-CalcEngine). 
Also, with this implementation, EnergyPlus will be able to use Window Calculation Engine as third-party component by simply inserting following lines in IDF:
```
WindowsCalculationEngine,
      ExternalWindowsModel;
```
In case some feature is still not supported by Windows Calculation Engine, code inside EnergyPlus will be used (ASHWAT is still not part of WCE).
For detailed description of algorithm please refer to Engineering Reference and also documentation hosted by LBNL:
https://windows.lbl.gov/sites/default/files/Downloads/WINDOW%20Technical%20Documentation.pdf

Current version of WCE used in EnergyPlus is 1.0.05.

vidanovic committed:
NFP for Windows-CalcEngine integration

vidanovic committed:
NFP update
Email comments added

vidanovic committed:
NFP update with conference call comments and suggestions

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
CMake integration of Windows-CalcEngine library

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

Commit:
Revert back CMake files (need to check why tests are not passing)

Some tests are not passing after CMake for Windows-CalcEngine was
implemented. There are no changes in EnergyPlus code

vidanovic committed:
Revert back CMake changes

vidanovic committed:
Separate module for window manager exterior added
Some routines refactored into surface (use from exterior window routines only)

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
Specular layers thermal algorithm implemented

vidanovic committed:
Thermal procedure in

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
Window model object added to IDD.
Thermal algorithm for specular layers.

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
Thermal part for specular layers with different gas mixtures

vidanovic committed:
Window-CalcEngine - Thermal engine integration

vidanovic committed:
Thermal calculation integration

vidanovic committed:
Diffuse optical calculations added

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
Fixed gcc and clang compilation errors

vidanovic committed:
Test hemispherical integration.

vidanovic committed:
Minor change in EnergyPlus window code. CMake compiler flag change in WindowsCalc-Engine.

vidanovic committed:
Base BSDF creators for different shade and specular materials.

vidanovic committed:
BSDF Integration routines added

vidanovic committed:
Fixed compilation error. Order of include files in Energy+ matters in some cases.

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
- Solar optical calculations - first pass (getting some results which needs to be tested)

vidanovic committed:
Update for Windows-CalcEngine renamed classes and files. Version is tagged.

vidanovic committed:
Missing solid layer for optical calculations.

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
Merged with 8.6 and added handling to shade

vidanovic committed:
Integration of new version of WCE for measured data.

vidanovic committed:
- BSDF IDF input fully functional with WCE (Winkelmann optical still missing)

vidanovic committed:
Consolidation of include for Windows-CalcEngine files. Now to use specific routines only one file needs to be included.

vidanovic committed:
Layer cells created in separate factory class.

vidanovic committed:
Fixes for optical calculations for Venetian Blind

vidanovic committed:
Redirect to new github version of WCE.

vidanovic committed:
Fixes some compilation errors in linux

vidanovic committed:
Some formatting and include statement cleanups.

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
Newer version of WCE. Some minor cleanup in CMake.

vidanovic committed:
Missing cmake files

vidanovic committed:
Merged with latest branch.

vidanovic committed:
gcc 4.9 tests on Linux
Some minor cleanup on Clang

vidanovic committed:
Fixes for layer(s) absorptance.

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
Minor unused variable cleanup.

vidanovic committed:
Removed unused variables.

vidanovic committed:
Some variables cleanup. Two additional functions in DataSurfaces.

vidanovic committed:
Airflow permeability layer is now created from input file.

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

# Conflicts:
#	src/EnergyPlus/HeatBalanceManager.cc

Conflict successfully merged

vidanovic committed:
Change for cmake integration test (folder path cannot be found)

vidanovic committed:
Reverted back RunSimulation.cmake

vidanovic committed:
Fix for Clang in use of shading devices.
Reorder insertion of WindowsCalc-Engine in cmake (possible interference with integration test)

vidanovic committed:
Revert back cmake inclusion of WCE to old place (caused crashing)
Renaming CommonIncludeDir to CommonWCEIncludeDir

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
Test for new cmake settings in WCE.

vidanovic committed:
config_compiler_and_settings renamed in WindowsCalc-Engine to avoid collision with E+ settings

vidanovic commented:
Need to add unit tests.

Myoldmopar commented:
@vidanovic This is looking much better.  

It will be a big shift on the build process, so I want to proceed down that path carefully.  Perhaps I could get @kbenne to peek at it if I say his name loud enough -- @KBENNE -- there, that ought to do it.

I see there are still no docs or unit tests.  We're running up against the IDD freeze, so I just wanted to get your feel for whether we'll be able to get this done in time.  Let me know, because if not, I will review other PRs that are more done, and this can drop into the main branch for the following release when it's done.

vidanovic commented:
@Myoldmopar I will be adding some test cases tomorrow. At this time I have no doc changes. How does it work? You need to merge it for IDD changes?

vidanovic commented:
@Myoldmopar However, I should update Input/Output reference with new IDF object. I should actually do that. Since that is minimal update, I will put it on schedule for tomorrow.

Myoldmopar commented:
@vidanovic glad to hear unit tests are en route.  Our dev schedule is set up such that we lock down the IDD "fully" about 3 iterations prior to release.  So any PR with any IDD change at all needs to go in by that time.  And before the PR must go in, it must be reviewed.  And usually some of the first things reviewers look for are the design doc, documentation, idd changes, and unit tests.  So that's why I noticed those couple things weren't there.

I'll wait to look again until tomorrow after I see your commits.  Thanks.

vidanovic committed:
- Input-Output reference documentation added.
- Several integration tests added.
- Reference to Windows-CalcEngine release Version_1.0.05 added

vidanovic commented:
@Myoldmopar I just added Intput/Output reference and several files for integration test. Unit test is done on our site (https://github.com/LBNL-ETA/Windows-CalcEngine).

Myoldmopar commented:
@vidanovic, thanks for adding the documentation.  

I don't really like the change to gcc 4.9 happening without any team discussion.  In fact, it has stopped the Linux CI from working on this branch.  We've held at 4.8 for a while and I'm not really interested in a hard compiler change this close to the release.  Can you change the code to get it back to 4.8?  Overall I don't mind a move to 4.9, but it can't really jeopardize the CI status and release.

As for the external project, I'm not really pumped about that either.  We've taken an approach to use a third_party folder to ensure we have absolute control over the source we are compiling.  I think we'd be more comfortable bringing in the window calc engine code into a third_party folder.  Can you bring that in here?

vidanovic commented:
@Myoldmopar We (team) actually talked about both things. For using external project in NFP and for gcc 4.9 I have asked at one of the meetings (about month ago) and had impression that is going to be ok. Problem is that gcc 4.8 have problem with lambdas which I extensively used in WCE. Let me talk to team here and I will come back to you.

mbadams5 commented:
@vidanovic Can you point to specific locations in your code that have trouble when compiling with gcc 4.8? I had mentioned on a call that i'd be willing to help investigate issues with compiling on gcc 4.8. It is my understand that gcc 4.8 has complete c++11 lambda support and c++14 lambda support in 4.9+, but I could be mistaken.

I also want to reiterate @Myoldmopar's point about the external project. I'm not a huge fan of it for various reasons and we already put third party code into a folder within EnergyPlus source code.

I will leave more specific comments as review comments in the code.

mbadams5 commented on src/EnergyPlus/DataSurfaces.cc:
```diff
@@ -86,6 +94,14 @@ namespace DataSurfaces {
 	using namespace DataPrecisionGlobals;
 	using namespace DataVectorTypes;
 	using namespace DataBSDFWindow;
+  using namespace DataHeatBalance;
```
You have numerous indention issues throughout the code. Looks like most are spaces when they should be tabs

vidanovic replied:
Then we should fix wiki documentation:
https://github.com/NREL/EnergyPlus/wiki/Coding-Standards-Part-1-Style
First it states "Style guidelines are not overly strict." and then later "Tabs are not allowed, and a mixture of tabs and spaces is strictly forbidden".

mbadams5 replied:
The style guidelines do probably need to be updated. Big things are to be consistent, use tabs, have one space between parentheses/other control mechanisms and the expressions/variables, try to avoid excessive `using namespace Foo;`, and try to declare variables close to their usage.

vidanovic replied:
Will do. Thanks again. Really appreciating all these comments. I was looking here at LBNL for someone to review my code, but everyone is busy. I also realized extensive use of shared_ptr (in WCE) while in most cases passing by reference would suffice.

mbadams5 commented on src/EnergyPlus/DataSurfaces.cc:
```diff
@@ -665,6 +685,184 @@ namespace DataSurfaces {
 			}
 		}
 
+    double SurfaceData::getInsideAirTemperature( const int t_SurfNum ) const
```
EnergyPlus uses `Real64` instead of `double` directly due to legacy translation. You should change all instances of `double` to `Real64` instead.

mbadams5 commented on src/EnergyPlus/HeatBalanceSurfaceManager.cc:
```diff
@@ -2432,7 +2461,31 @@ namespace HeatBalanceSurfaceManager {
+                  double Theta = incomingAngle.first;
+                  double Phi = incomingAngle.second;
+
+                  std::shared_ptr< CMultiLayerScattered > aLayer =
```
Why does this need to be a `std::shared_ptr`? It doesn't seem to actually be "shared" anywhere and just adds overhead.

mbadams5 commented on src/EnergyPlus/SolarShading.cc:
```diff
@@ -7017,7 +7035,330 @@ namespace SolarShading {
+
+        int ConstrNum = Surface( SurfNum2 ).Construction;
+        if( window.ShadedConstruction > 0 ) ConstrNum = window.ShadedConstruction;
+        auto aLayer = CWindowConstructionsSimplified::instance().
```
Again this `auto` is hiding a `shared_ptr` that isn't actually "shared".

vidanovic commented:
@mbadams5 OK. Lets talk about this on one of the meetings. Seems like I am not going to make it to this release.
Related to lambdas. Can you see this link:
https://github.com/LBNL-ETA/Windows-CalcEngine/branches

?

mbadams5 commented on src/EnergyPlus/WindowManager.hh:
```diff
@@ -47,7 +47,10 @@
+#include <memory>
+
 // ObjexxFCL Headers
+#include <ObjexxFCL/Array1D.hh>
```
Why do you need this include in the header file? If it is needed in the implementation file, it should go there.

mbadams5 commented on src/EnergyPlus/WindowManagerExteriorData.cc:
```diff
@@ -0,0 +1,346 @@
+
+namespace EnergyPlus {
+
+  using namespace std;
```
No. Please don't use `using namespace std;` it is bad enough all the other using namespaces.

mbadams5 commented on src/EnergyPlus/WindowManagerExteriorData.cc:
```diff
@@ -0,0 +1,346 @@
+
+#include <assert.h>
+#define _USE_MATH_DEFINES
+#include <math.h>
```
Why `#include <math.h>` and not `#include <cmath>`?

mbadams5 commented on src/EnergyPlus/WindowManagerExteriorData.cc:
```diff
@@ -0,0 +1,346 @@
+// in binary and source code form.
+
+#include <assert.h>
+#define _USE_MATH_DEFINES
```
What is this `#define` used for?

vidanovic replied:
I use it for PI (constant name is M_PI).

mbadams5 replied:
[EnergyPlus already has a PI constant](https://github.com/NREL/EnergyPlus/blob/17d4d9083dfca8ff7de24a22938d61e53ace0fb7/src/EnergyPlus/DataGlobals.cc#L110) and [M_PI is not part of c/c++ standard...](https://msdn.microsoft.com/en-us/library/4hwaceh6.aspx). But i'll let others weigh in on using M_PI (and potentially other math constants in cmath).

vidanovic replied:
That is actually interesting problem. WCE will be used in other tools (WINDOW) and we will have commonly shared code between EnergyPlus and WCE. For example, I have classes that deal with matrices and I am sure EnergyPlus have them too. But for me to use it, I would need to pull that out and keep it in synch. Do you think that it would be a good idea to pull some of those common routines out? (or even better to use some third party component, boost, etc.)

mbadams5 commented on src/EnergyPlus/WindowManagerExteriorData.cc:
```diff
@@ -0,0 +1,346 @@
+    }
+
+    pair< double, double > getSunWCEAngles( const int t_SurfNum, const BSDFHemisphere t_Direction ) {
+      std::pair< double, double > Angles;
```
This doesn't seem to be used.

mbadams5 commented on src/EnergyPlus/WindowManagerExteriorOptical.cc:
```diff
@@ -0,0 +1,651 @@
+
+namespace EnergyPlus {
+
+  using namespace std;
```
Don't use `using namespace std;`

mbadams5 commented on src/EnergyPlus/WindowManagerExteriorThermal.cc:
```diff
@@ -0,0 +1,663 @@
+
+namespace EnergyPlus {
+
+  using namespace std;
```
Don't use `using namespace std;`.

mbadams5 commented on src/EnergyPlus/WindowManagerExteriorThermal.cc:
```diff
@@ -0,0 +1,663 @@
+    }
+
+    /////////////////////////////////////////////////////////////////////////////////////////
+    MaterialProperties* CWCEHeatTransferFactory::getLayerMaterial( int const t_Index ) const {
```
You use `shared_ptr` everywhere else but here you are returning a raw pointer, I assume because it is pointing to an EnergyPlus data structure. I see it is `private` in the class so i'm less concerned about this.

mbadams5 commented on src/EnergyPlus/WindowModel.cc:
```diff
@@ -0,0 +1,95 @@
+
+#include "WindowModel.hh"
+
+using namespace std;
```
Don't use `using namespace std;`

mbadams5 commented on src/EnergyPlus/WindowModel.cc:
```diff
@@ -0,0 +1,95 @@
+    }
+  
+
+    shared_ptr< CWindowModel > CWindowModel::WindowModelFactory( string objectName ) {
```
Should pass by reference. `std::string const & objectName`

mbadams5 commented on src/EnergyPlus/WindowModel.hh:
```diff
@@ -0,0 +1,62 @@
+#define WindowModel_H
+
+#include <memory>
+#include <iostream>
```
Why do you need to include `#include <iostream>` here? it doesn't seem like you are using any IO within the header file.

vidanovic replied:
@mbadams5 I have done requested changes. Can you please review them?

mbadams5 reviewed (changes requested):
You make immense use of `std::shared_ptr` and it is hard to tell if you actually need and use the sharing semantics of it or not. It seems like in most places you do not. Using `shared_ptr` has an increased overhead over `unique_ptr` or regular objects. It is also hard to see from the code (and without the linux CI running) what the performance impacts of using `shared_ptr` really is for these changes. It could be that it is minimal but without profiling it is hard to tell. Anyways, i'm curious to hear your thoughts on why you chose to use `shared_ptr` through almost all your code.

mbadams5 commented:
@vidanovic Yes I can see that link. Can you point to any specific spots in your code that are failing with gcc48 please? We can also have a call to chat about this or another form of communication (slack, etc).

vidanovic commented:
@mbadams5 Thanks for the comments.
Yes, WCE is using shared_ptr(s) a lot. For most of the cases I have found it necessary since I do share objects between calculations. I am sure that some of them can be used as unique_ptr or even just plain objects.
I use have hangouts here. Can also use Zoom to make conf calls.

mbadams5 commented:
I see [this line is failing in your travis build for example.](https://github.com/LBNL-ETA/Windows-CalcEngine/blob/master/src/Common/src/Polynom.cpp#L62) You do not need to capture the outer scope by reference for this lambda and should be removed. That should fix your build failure for that specific lambda. [For the lambda a few lines down,](https://github.com/LBNL-ETA/Windows-CalcEngine/blob/master/src/Common/src/Polynom.cpp#L73-L74) instead of capturing the full outer scope by reference, I would just pass the required outer variable by value instead. So I think this is accomplishable to get your code running on gcc4.8 by better use of the lambda semantics.

We can chat more.

vidanovic commented:
@mbadams5 Thanks! I have sent you the invite.

vidanovic commented:
@Myoldmopar and @mbadams5 Just had short meeting with team here. I got reminded that this project is guinea pig for external project. Beside other things, idea was to see how will project behave if we do something like that (gtest is example of external project as well). There are certain const and pros for sure.
Lets discuss this on Wednesday's meeting.

mbadams5 commented:
@vidanovic I just want to quickly comment that within EnergyPlus ALL third party code, including gtest, eigen, sqlite, kiva (!!), etc are fully within the third_party folder. Some of these could have used external_project, but were put into third_party directly. Kiva is a prime example of a project that could have easily used external_project, but was put into third_party.

Anyways, we can chat more about this in the next meeting.

vidanovic commented:
@mbadams5 and they should be hosted separately if they are to be used by other programs as well. One idea for WindowsCalc-Engine is to be used by LBNL WINDOW and possible as standalone engine. I was also told that Amir liked that idea and wanted to see how it works. Hopefully he will be on the meeting as well.
BTW, just sent you invitation on hangouts. Don't have to chat now...just sent invitation so we can do it in the future.

Best,
Simon (Dragan) :)

jasondegraw commented:
@mbadams5 @vidanovic Maybe you've already discussed this, but have we considered git submodules instead of CMake external projects?

vidanovic commented:
@jasondegraw @mbadams5 We actually had similar discussion when NFP was out. Unfortunately I did not put who was on the meeting. I cannot recall discussing for other modules but I was not regular on E+ meetings for some time.

mbadams5 commented:
@jasondegraw .... 👎 🚫 

And nothing is stopping these external programs from being hosted elsewhere. All the ones I listed ARE hosted elsewhere. We just include a version of it at a specific point in time. So functionally it is similar to how external_project works, but without the external link.

jasondegraw commented:
@mbadams5 @vidanovic Seems like that is not a popular idea. I've never used submodules myself, but my experience with external projects has been mainly bad. So I was wondering if submodules would be a better path.

Myoldmopar commented:
I'm in favor of paths that keep us in control of the code we are compiling and ultimately releasing to the public.  With an external project, there is nothing stopping us from being unable to produce a full installation package.  And if we rely on a specific tag, there is nothing stopping that tag from being deleted and recreated later for a different SHA.  Causing chaos.  For me, it's third_party or bust.

vidanovic commented:
@Myoldmopar I assume that you are not allowing people to change third_party in EnergyPlus directly? Please, keep in mind that keeping separate code is exactly what caused chaos in windows related routines and we are supposed to explain customers why are they getting results (especially with AERC coming up). That is why this approach is coming from lbnl.
Also, I am not the only one that is suppose to change decision about external vs third_party, otherwise it would be easier. But if decision was made within last year, then I suppose we have to adopt as well. Finally, we are now in github world and maybe things will be easier. I am would like to hear what problems did you have so we don't repeat them here at lbl? I can also use it as argument for third_party.

Myoldmopar commented:
Here is the situation I worry about, @vidanovic.  Let's say you leave lbnl in 4 weeks, right before E+ release.  And the next person to maintain this library decides that day to just make the repo private, or delete the git tag `Version_1.0.05`, or rename the git repo.  Then we kick off our CI builds to do the release of EnergyPlus 8.8, and boom.  We cannot build our releases anymore, because of something totally out of our control.

If we keep just one revision, the revision we want (tag `Version_1.0.05`) in a directory in the third_party folder, then we won't have this potential problem.  Our own repo, that we clone and maintain, has enough code to generate all libraries, documentation, and binaries necessary to create installer packages.

As you make changes to the Window library, just drop those into the third_party folder as updates.  It's very seamless, and is how Objexx, SQL, and Kiva all work for us.  It makes our repo slightly larger, but ensuring that we can build a package with the code we control is ultimately very important to me.

vidanovic commented:
@Myoldmopar That is good reason. No one can guarantee that tag or repo will not be deleted (I think that actually happened with Node npn for javascript). How do you update kiva? Are you doing it mainly on kiva web site or you do updates on EnergyPlus git as well and then take it back to kiva?

jasondegraw commented:
@vidanovic Since you asked about problems, here's a quick summary of mine. Most, if not all, of my bad experiences with ExternalProjects involve git or GitHub, so that doesn't factor in for me. It could be that it works better when not using Visual Studio, but I've had trouble with ExternalProjects not being built at the right time (rebuilt when there is no need). I have experienced the issue @Myoldmopar is talking about when the build can't access the URL of the project (network goes down, no build for me). Maybe I'm just bad at using it, but all in all, pretty frustrating. The only unqualified success I've had with ExternalProject was with a header-only library that I stashed as tarball in the repo (please don't ask why). Some would argue that was not a success.

Myoldmopar commented:
It's not really any different than I suggested.  Just think of the third_party folder as the external project, but instead of getting the project from the Git repo every time, it already has it within the NREL/EnergyPlus repo in the third_party folder.  It can be the exact same code from the original third_party repo every time.

So with Kiva, @nealkruis can work away on Kiva enhancements to his heart's content.  Then, one day, he decides it'd be nice to get these changes back in E+'s version of Kiva.  In an ExternalProject approach, he would specify an updated Git Tag over here in E+ and issue that as a PR.  In the third_party approach, he'd simply copy/paste the entire third_party/kiva folder with his new version, and we'd review the diffs to make sure the API isn't broken and E+ is still happy.  Then merge it.  And done.

vidanovic commented:
@jasondegraw Thank you guys! It is also important for me to know before we set up THERM and WINDOW world (and most likely OPTICS and THERMM later) so it is good to share experience. We are currently heading towards git submodules, which seems to me like really bad idea. Any experience with this?

Myoldmopar commented:
I tried using them just to learn about them quite a while back.  You'll run into the same problem with not being able to get to a remote source if they change it.  When you first clone a repo with submodules, it won't go out and grab the latest, so it adds to more steps to the workflow.  

We'll use third_party for the foreseeable future.

vidanovic committed:
Spaces to tabs

vidanovic committed:
- dobule to Real64 (keeping legacy use of Real64).
- Window model factories are now unique_ptr.
- Some additional formatting issues.

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

# Conflicts:
#	src/EnergyPlus/DataSurfaces.cc
#	src/EnergyPlus/DataSurfaces.hh
#	src/EnergyPlus/WindowManager.cc
#	testfiles/CMakeLists.txt

vidanovic committed:
Fix for unique_ptr creation (test failures)

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
Windows-CalcEngine into third_party software

vidanovic committed:
- Removed std:move from several places
- included string for GasData file

vidanovic committed:
BSDF layer factory removed.

vidanovic committed:
WindowCalc-Engine unit testing now compiled with EnergyPlus.

vidanovic committed:
Merge branch '#2520943-WindowsCalc-Engine_Integration' of https://github.com/NREL/EnergyPlus into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
Removed shared_ptr use in push_layer routine

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
Merge with develop and revert back decent_ci.yaml to gcc 4.8

vidanovic committed:
Update to test file for version 8.8 (it still needs input for Windows-CalcEngine)

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
License text at the top of the file updated.

vidanovic committed:
removed M_PI (from <math.h>) <assert.h> replaced with <cassert>

vidanovic committed:
License update for several files

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic commented:
@mbadams5 I did update code. Also put WCE into third_part. I even removed some of shared_ptr, but not all of them since it will take some time. If time allows, can you please review the code again and see if you see some other problems? Thanks.

vidanovic committed:
License text modified (paragraph removed)

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

# Conflicts:
#	src/EnergyPlus/WindowComplexManager.cc

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

nrel-bot-2 commented:
@vidanovic @lgentile it has been 14 days since this pull request was last updated.

mjwitte commented:
@vidanovic Waiting for build warnings and test failures to be fixed before reviewing further.

vidanovic commented:
Yes, Im working on this. Hope to catch some time on Monday so that I can post new code.

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
Formatting fixes.
New version of WCE.

vidanovic committed:
back_inserter not recognized by Visual Studio 14 (include added)

vidanovic committed:
Update to license text and few cpp_check changes

vidanovic commented:
@mjwitte @Myoldmopar I need help with these errors:
https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-2013128b7404cba9ae9b570277fb381be9e2cce0-x86_64-Linux-Ubuntu-14.04-gcc-4.8.html
Ignore first 4 since they are new, but why rest of them keep failing? From the message I would say it is AUD difference, which is actually expected since I have new IDD object.
Thanks.

mjwitte commented:
@vidanovic When you see "bucket" or "S3" errors, it's a CI problem, nothing to do with your code, so you can ignore those.  You can delete these results and wait for them to re-run, or just wait until your next commit.

vidanovic commented:
Wonderful. I have small change, so will recommit again.

vidanovic committed:
Another cpp_check update

vidanovic committed:
Disabled gcc message for enable_shared_from_this copy constructor

vidanovic committed:
using namespace std removed from WindowsCalc-Engine library

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
removed USE_MATH_DEFINES for use with M_PI. Constant PI is now defined in local library in common module.
Fixed usage of std::abs (crash on clang)

vidanovic committed:
Remove usage of M_PI from main window modules in E+ (caused compilation error in VS2015)

vidanovic commented:
@mjwitte @Myoldmopar I have cleaned up everything else but these messages:
https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-a5a3d0fe6439f4201682103a29a705574ef711fe-x86_64-Linux-Ubuntu-14.04-gcc-4.8.html
My only worry is that something else in the code is causing them, because every time I upload some change I get them. Maybe I should try to remove my tests and see if program will crash again? However, my tests do work locally.

mjwitte commented:
@vidanovic I can't see the error file on the CI, but the WCE files are likely failing because they do not have a location object in them.  CI runs design-day only simulations on the test files, so there must be a location object in the idf.

vidanovic committed:
Location object added to WCE test files

vidanovic commented:
@Myoldmopar Can I get err files from ubuntu run(s)? Not sure what is going wrong on gcc machine. Thanks.

mjwitte commented:
@vidanovic The WCE files are failing on all platforms.  Ran locally on Windows, without a weather file, and with simulationcontrol changed to do sizing periods= yes, no to weather file periods (like they run on CI).
   ** Warning ** ManageSimulation: SizingPeriod:* were requested in SimulationControl but no SizingPeriod:* objects in input.
Sorry I didn't notice that before.  So, these all need a couple of SizingPeriod:DesignDay objects also.

mjwitte commented:
@vidanovic And these should be using a shorter timestep.
```
   ** Warning ** Timestep: Requested number (1) is less than the suggested minimum of 4.
```

mjwitte commented:
And there are some other warnings that should be cleaned up.  These are from WCE_Diffuse_Shade.idf.  I haven't run the others.  
```
   ** Warning ** ProcessScheduleInput: Schedule:Compact="ZONE CONTROL TYPE SCHED", Schedule Type Limits Name="NUMBER" not found -- will not be validated
   ** Warning ** ProcessScheduleInput: Schedule:Compact="SETPOINT COOLING", Schedule Type Limits Name="NUMBER" not found -- will not be validated
   ** Warning ** ProcessScheduleInput: Schedule:Compact="SETPOINT HEATING", Schedule Type Limits Name="NUMBER" not found -- will not be validated
   ** Warning ** GetSurfaceData: WindowProperty:ShadingControl: "SHDSYS_CONTROL" is not referenced by any window.
   ** Warning ** CheckUsedConstructions: There are 1 nominally unused constructions in input.
```
Also, these need the standard comment blocks at the top of the file so the automated scripts can populate the example file indexes.  See other example files, such as 5ZoneAirCooled for format.  In a full install, some of the comment information is used to populate ExampleFiles.html and ExampleFiles-ObjectsLink.html.

nrel-bot-3 commented:
@vidanovic @lgentile it has been 14 days since this pull request was last updated.

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
Merge and fix for test files.

vidanovic committed:
Fix for woven shade input file

vidanovic committed:
Removed problematic test file

vidanovic commented:
@mjwitte Just did additional cleaning. Had errors on gcc 4.8, but after rebuilding it errors are gone. Let me know how this look now? Thanks.

nrel-bot-3 commented:
@vidanovic @lgentile it has been 14 days since this pull request was last updated.

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
jsoncpp was merged in by mistake

energyplus-style-bot committed:
Merge commit '9768414a04e06361a0c0afedcb0a8b89e5a12b10' into #2520943-WindowsCalc-Engine_Integration

energyplus-style-bot committed:
Merge commit '61b09a18629a347273e5d099c3f2e3acbcbc0da4' into #2520943-WindowsCalc-Engine_Integration

energyplus-style-bot committed:
apply clang-format

energyplus-style-bot committed:
Merge commit 'b6f573e2e9a42174cad2383d458fefad5b62a211' into #2520943-WindowsCalc-Engine_Integration

energyplus-style-bot committed:
apply clang-tidy

energyplus-style-bot committed:
apply clang-tidy

energyplus-style-bot committed:
Fix issue from reformatting with DataSurfaces

vidanovic committed:
WindowManager.cc did not have WindowManager.hh included.

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
Update to licensing year (2018).

nrel-bot-3 commented:
@vidanovic @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@vidanovic @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-2 commented:
@vidanovic @lgentile it has been 14 days since this pull request was last updated.

nrel-bot-3 commented:
@vidanovic @lgentile it has been 14 days since this pull request was last updated.

Myoldmopar commented:
Alright, this is on my list @vidanovic.  I'll be taking a look shortly.  @mbadams5 is your rejected review still valid?  I'm not sure how old your review is but I don't want to duplicate anything.

vidanovic commented:
Thanks. Yes, @mbadams5 still did not do second review. I did address all his comments expect shared pointers since that would require some major rework and it is not critical at the moment.

mbadams5 commented on src/EnergyPlus/SolarShading.cc:
```diff
@@ -7522,6 +7532,317 @@ namespace SolarShading {
+                Real64 Tdiff = aLayer->getPropertySimple(PropertySimple::T, Side::Front, Scattering::DiffuseDiffuse, Theta, Phi);
+                Construct(ConstrNum).TransDiff = Tdiff;
+                Real64 DSZoneWin = window.SkySolarInc * Tdiff * Surface(SurfNum2).Area;
+                (DifSolarRad != 0) ? DSZoneWin /= DifSolarRad : DSZoneWin /= 1e-8;
```
This is an odd way to write this conditional and confusing at first glance. This won't hold up the PR from my end but if you need to address other changes then I'd change this as well.
```
DSZoneWin /= ((DifSolarRad != 0) ? DifSolarRad : 1e-8);
```

mbadams5 commented on src/EnergyPlus/WindowManagerExteriorData.cc:
```diff
@@ -0,0 +1,337 @@
+    ///////////////////////////////////////////////////////////////////////////////
+    CWindowConstructionsSimplified &CWindowConstructionsSimplified::instance()
+    {
+        static CWindowConstructionsSimplified p_inst;
```
Is this used in unit tests? Does it need to be "reset" between unit tests? The way it is written now, it can not be "reset" between different unit tests.

mbadams5 commented on src/EnergyPlus/WindowManagerExteriorData.cc:
```diff
@@ -0,0 +1,337 @@
+    void
+    CWindowConstructionsSimplified::pushLayer(WavelengthRange const t_Range, int const t_ConstrNum, std::shared_ptr<CScatteringLayer> const &t_Layer)
+    {
+        Layers_Map &aMap = *m_Layers.at(t_Range);
```
You are already using `.at()`, why not make it `m_Layers->at(t_Range);` instead? This is more of a style comment and not critical for merge.

mbadams5 commented on src/EnergyPlus/WindowManagerExteriorData.cc:
```diff
@@ -0,0 +1,337 @@
+        if (it == m_Equivalent.end()) {
+            // Layer was not requested before. Need to create it now.
+            // shared_ptr< vector< double > > commonWl = getCommonWavelengths( t_Range, t_ConstrNum );
+            IGU_Layers iguLayers = *getLayers(t_Range, t_ConstrNum);
```
From a code understanding point of view, I don't like dereferencing a return value from a function like this. However, it isn't critical for merge.

mbadams5 commented on src/EnergyPlus/WindowManagerExteriorData.cc:
```diff
@@ -0,0 +1,337 @@
+        Layers_Map aMap = *m_Layers.at(t_Range);
+        auto it = aMap.find(t_ConstrNum);
+        if (it == aMap.end()) {
+            throw std::runtime_error("Incorrect construction selection.");
```
Throwing an exception here will abort the simulation immediately (assuming you don't have a try/catch somewhere else). Do you want simulation to stop or continue for a little bit and stop later?

mbadams5 commented on src/EnergyPlus/WindowManagerExteriorOptical.cc:
```diff
@@ -0,0 +1,645 @@
+                    if (material.Group != WindowGas && material.Group != WindowGasMixture && material.Group != ComplexWindowGap &&
+                        material.Group != ComplexWindowShade) {
+                        auto aMaterial = std::make_shared<MaterialProperties>();
+                        *aMaterial = material;
```
I'm a little uncertain of this behavior. It reads like you are assigning the `material` pointer to the created `shared_ptr`. In which case, what happens when `aMaterial` has 0 references and cleans itself up? Does it delete the `material` object? I'm not quite sure, I need to research a little bit more.

vidanovic replied:
@mbadams5 it calls operator=() to assign data from one object to another

mbadams5 replied:
I understand that it calls the assignment operator, but what I meant is who owns that data at that point. For example, I know unique_ptr expects to completely own the underlying data (pointer) and will clean up that resource when it falls out of scope. What I am unsure about is whether shared_ptr behaves similarly when the reference count goes to 0, I believe it does though.

Additionally if the Material() isn't created with `new` and the shared_ptr falls out of scope with 0 references, it will probably throw an exception since it doesn't know how to `delete` something not created by `new`.

vidanovic replied:
@mbadams5 @jasondegraw You are talking of this issue? I think it is ok, but to be on safe side I think shared_ptr can be removed here. Will take a look later today if I can remove it. Should not be the problem.

mbadams5 commented on src/EnergyPlus/WindowManagerExteriorOptical.hh:
```diff
@@ -0,0 +1,336 @@
+                                     const FenestrationCommon::WavelengthRange t_Range);
+
+    private:
+        void init();
```
This should probably be:
```
virtual void init() override;
```
But this is just being more explicit than what you currently have, which compiles fine.
This comment applies to all instances of virtual methods in your various factory classes.

mbadams5 commented on src/EnergyPlus/WindowModel.hh:
```diff
@@ -0,0 +1,118 @@
+    T StringToEnum(const std::string &value)
+    {
+        auto iValue = m_Map.find(value);
+        if (iValue == m_Map.end()) throw std::runtime_error("Incorrect enumerator assigned.");
```
Like my other comment, throwing an exception will cause the simulation to end abruptly, if that is the desired outcome.

mbadams5 commented on third_party/Windows-CalcEngine/cmake/WCECompilerFlags.cmake:
```diff
@@ -0,0 +1,50 @@
+# Make sure expat is compiled as a static library
+# ADD_DEFINITIONS("-DXML_STATIC")
+
+IF ( CMAKE_COMPILER_IS_GNUCXX OR "x${CMAKE_CXX_COMPILER_ID}" STREQUAL "xClang" ) # g++/Clang
```
Is it supposed to be `xClang` or just `Clang`?

mbadams5 reviewed (commented):
I made some style comments and other general question comments. Nothing that would hold up this PR but if you make other changes then it would be nice to address these comments as well.

Do you have EnergyPlus unit tests (unless I missed them)? I see many WCE unit tests.

I also want to voice again that i'm not a fan of pervasive usage of shared_ptr. They can quickly and easily hide performance degradation issues. However, I understand it would require significant effort to refactor the code to limit the usage of shared_ptrs.

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
Merge with base and some code cleanup

vidanovic commented:
@mbadams5 thanks. Those are useful comments.
@Myoldmopar I will merge with develop and address some of the issues.

jasondegraw commented:
@Myoldmopar @vidanovic I would have commented sooner, but I've been seeing intermittent angry unicorns since the iteration call. I'm going to second @mbadams5 on the use of shared pointers, and I'm not a fan of the use of exceptions in the main code. I thought the consensus was to keep the main code exception free.

vidanovic commented:
@jasondegraw I can remove exceptions in main code (have couple of them). However, I still keep them in component. I assume prefference is ShowSevereError("Message here.") or ShowContinueError("Entered in " + cCurrentModuleObject + '=' + cAlphaArgs(1))

vidanovic committed:
- Error message handling is using EnergyPlus message system.
- Unit test for WCE_Reader

jasondegraw commented:
@vidanovic Yes, `ShowSevereError` if execution is to continue but stop later (for example during input) or you can use `ShowFatalError` if you want the simulation aborted.

vidanovic commented:
@jasondegraw I have put this to catch error from WCE:

https://github.com/NREL/EnergyPlus/tree/%232520943-WindowsCalc-Engine_Integration/src/EnergyPlus/WindowManagerExteriorThermal.cc#L121-127

That would do it.

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
Resetting tests (dont know how to do it online)

jasondegraw commented:
@vidanovic I'd personally prefer to have those wrapped up in a way that we don't see that in the main code, but that's good enough for now.

vidanovic commented:
@jasondegraw Not sure that I understand. Maybe we can do some goto meeting or maybe some document somewhere is explaining how to do it?

jasondegraw commented:
@vidanovic I don't think it's worth holding things up over this, let's just leave it as is. It's a lot less important than the ownership issue that @mbadams5 pointed out.

vidanovic committed:
Removed shared_ptr from material

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
xClang to Clang

vidanovic committed:
- Material no longer needs assert for nullptr because it is no longer pointer
- Clearing up shared pointers from some maps

vidanovic committed:
Updated singleton for CWindowConstructionsSimplified

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
static member was not initialized in cpp file

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
Removed forward class declaration in singleton

Commit:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

Myoldmopar committed:
Apply clang-format to WindowManagerExteriorData.* files

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into #2520943-WindowsCalc-Engine_Integration

Myoldmopar commented:
@vidanovic CI was [complaining](https://github.com/NREL/EnergyPlus/pull/6247#commitcomment-29711152) that a couple of files weren't properly formatted with our new style.  I applied the style, merged develop, and then pushed the code back up just now.  Should be 100% green now.

vidanovic commented:
@Myoldmopar thanks.

Myoldmopar commented on idd/Energy+.idd.in:
```diff
@@ -8701,6 +8701,15 @@ WindowThermalModel:Params,
+       
+WindowsCalculationEngine,
+       \memo This object is used to switch algorithms between built in windows engine 
+       \memo and windows engine as exterior component
```
This sentence is pretty vague, if any other last minute code changes are needed, I'd like this to be explained a little better.

vidanovic replied:
You are right. This can be better explained. Will do.

Myoldmopar commented on src/EnergyPlus/DataSurfaces.cc:
```diff
@@ -677,6 +694,191 @@ namespace DataSurfaces {
+        return value;
+    }
+
+    Real64 SurfaceData::getSWIncident(const int t_SurfNum)
```
Should these little `get` functions be `inline`d?  Maybe the compiler is already inlining them.

vidanovic replied:
I think static is inline by default. Not sure.
BTW, I was extracting some common routines out. Like Short-Wave incident incoming to surface is calculated on so many places using same equation:

QRadSWOutIncident(t_SurfNum) + QS(Surface(t_SurfNum).Zone)

I was wondering maybe we can use clean E+ even more by identifying such functions.

Myoldmopar replied:
As long as they are inlined then I'm good with it.  I just don't want accessor functions to add any overhead.

Myoldmopar commented on src/EnergyPlus/WindowManagerExteriorData.hh:
```diff
@@ -0,0 +1,187 @@
+        static std::shared_ptr<FenestrationCommon::CSeries> getDefaultVisiblePhotopicResponse();
+    };
+
+    // ///////////////////////////////////////////////////////////////////////////////
```
Bulky comments left in?

Myoldmopar commented on third_party/Windows-CalcEngine/CMakeLists.txt.in:
```diff
@@ -0,0 +1,18 @@
+# find_package(Threads REQUIRED)
```
We aren't using this for normal development right?  I don't want to have to go out and have the ExternalProject of WCE testing when I'm doing my own development.  I'm already planning on turning off the Objexx testing for our builds, please confirm this is disabled by default.

vidanovic replied:
@Myoldmopar Just one clarification here. If you are talking of WCE unit testing this is off by default. However, if you check BUILD_TESTING flag in CMake then it will build unit tests for WCE. This is actually how other components are working.

Myoldmopar replied:
Yeah I know this is how things have worked, but I realized it's taking time to build when I don't care about actually running the tests for our third_party libraries.  Of course it's necessary for you to have them, and we should be able to run them whenever we want.  But very soon turning on BUILD_TESTING for EnergyPlus will not automatically turn on the testing for Objexx unit tests, or Kiva tests and it shouldn't turn on the unit tests for WCE either.  Those should be available through a secondary testing switch that defaults to OFF.  While you are making your final changes please include support for this.

vidanovic replied:
Should I add flag BUILD_WCE_TESTING? Is that how other components will be implemented as well?

Myoldmopar replied:
Yes please, that's perfect.  Inside the IF(BUILD_TESTING) block, add the new BUILD_WCE_TESTING flag, default to OFF.  That will make it available to an interested developer, but not waste our resources building it for every single commit.  Thanks!

vidanovic replied:
Done

Myoldmopar commented on tst/EnergyPlus/unit/WCE_Reader.unit.cc:
```diff
@@ -0,0 +1,185 @@
+
+// EnergyPlus Headers
+#include <DataIPShortCuts.hh>
+/*#include <ConvectionCoefficients.hh>
```
:-1: to C style commenting...just delete these includes if they aren't needed.

vidanovic replied:
Forgot to remove that. Will do.

Myoldmopar commented on tst/EnergyPlus/unit/WCE_Reader.unit.cc:
```diff
@@ -0,0 +1,185 @@
+    const auto Rback = solarLayer->getPropertySimple(FenestrationCommon::PropertySimple::R, FenestrationCommon::Side::Back,
+        FenestrationCommon::Scattering::DirectDirect);
+    EXPECT_NEAR(0.073682, Rback, 1e-6);
+}
```
Succinct, which I don't mind, but is it sufficiently covering the E+ implementation?  It seems like you have plenty of coverage of WCE itself in there, but I am more concerned about the E+ mechanics and making sure we hit as many corner cases as possible.

vidanovic replied:
This part should call not just reading but also thermal and optical part of WCE calculation within EnergyPlus. Yes, I am pulling numbers out, but these numbers are obtained by calling EnergyPlus routines. Will check this again on Monday.

Myoldmopar reviewed (commented):
A few comments here and there.  This is really really close.  CI looks 100% happy, so that's great news.  Please respond to my comments and make any final code changes to address them.  I'm happy with this going in once my comments are addressed through code changes or just leaving it as is and justifying it.

Myoldmopar commented:
(And if you make the code changes, make sure to pull develop in because a merge I did just now will cause diffs in 1 file)

vidanovic commented:
@Myoldmopar I always try not to forget to merge. Will do.

vidanovic commented:
I should also not forget to update IDF files to 9.0.

vidanovic committed:
Additional test cases added.
Minor fix for wce diffuse shade updated.
Update to WCE file version.

vidanovic committed:
Merge branch 'develop' into #2520943-WindowsCalc-Engine_Integration

vidanovic committed:
Windows-CalcEngine testing is now off by default even when EnergyPlus unit test is on.

vidanovic committed:
WCE_Diffuse_Shade correct file inserted.

Myoldmopar commented:
Looking very promising with the last two CI reports.  If CI is clean here I'll merge.

Myoldmopar commented:
(And by _clean_, there may be an ERR diff popping in to the Linux GCC build from #6816 that I just merged)

mjwitte commented:
@vidanovic Please edit the top comment in the pull request with a user perspective description of the new feature and other changes.

vidanovic commented:
@mjwitte Is there a way to edit it without committing entire branch? I see documentation to do it locally, before any commit, but not after push.

vidanovic commented:
@Myoldmopar Let me just change top comment of commit so that user can understand better. Will do that in few hours.

rraustad commented:
Click the pencil at the top right corner of the pull request overview. Hold you cursor over it and it says "edit". Click and edit as needed.

mjwitte commented:
Edit it directly in github on the pull request page, like @rraustad said.

vidanovic commented:
@rraustad @mjwitte Thanks. Done. For some reason I was reading "Please edit the top comment.." as "Please edit the top commit...". Should have drink coffee first :)
@Myoldmopar I have edited top comment. If all test pass, then I am ready. Let me know if you have any comments.

Myoldmopar commented:
:+1: Thanks to all who have contributed to this.  Linux will be done soon and this will go in.

Myoldmopar commented:
OK, I have no idea why Linux builds are taking their time.  I know they got a bit bogged down, but they are churning out builds now just fine, and I would've thought they'd be done here by now.  I just built on Linux and it built quick with BUILD_TESTING=ON (thanks for disabling the third party tests @vidanovic).  I'm going to go ahead and merge this in and we can move on.  

 :balloon:  :balloon:  :balloon:  :balloon: Thanks again all.  :balloon:  :balloon:  :balloon:  :balloon:

vidanovic commented:
@Myoldmopar I was wondering very same thing. Thanks everyone for help.
